source/particles/property_pool.cc

```diff
@@ -56,27 +96,54 @@ namespace Particles
 
 
   void
-  PropertyPool::deallocate_properties_array(Handle handle)
+  PropertyPool::deallocate_properties_array(Handle &handle)
   {
-    Assert(currently_open_handles.count(handle) == 1, ExcInternalError());
-    currently_open_handles.erase(handle);
-    delete[] handle;
+    Assert(
+      handle != invalid_handle,
+      ExcMessage(
+        "This handle is invalid and cannot be deallocated. This can happen if the "
+        "handle was deallocated already before calling this function."));
+    currently_available_handles.push_back(handle);
+    handle = invalid_handle;
+
+    // If this was the last handle, resize containers to 0.
+    // This guarantees that all future properties
+    // are allocated in a sorted way without requiring reallocation.
+    if (currently_available_handles.size() * n_properties == properties.size())
+      {
+        currently_available_handles.clear();
+        properties.clear();
+      }
   }
 
 
 
   ArrayView<double>
   PropertyPool::get_properties(const Handle handle)
```

Comment:
As this is a very small function, would you mind moving the implementation to the header file so it can be inlined?